src/box/lua/func_adapter.c

```diff
@@ -0,0 +1,280 @@
+lua_func_adapter_begin(struct func_adapter *base,
+		       struct func_adapter_ctx *base_ctx)
+{
+	struct lua_func_adapter *func = (struct lua_func_adapter *)base;
+	struct lua_func_adapter_ctx *ctx =
+		(struct lua_func_adapter_ctx *)base_ctx;
+	if (fiber()->storage.lua.stack == NULL) {
+		ctx->L = luaT_newthread(tarantool_L);
+		if (ctx->L == NULL)
+			panic("Cannot create Lua thread");
+		ctx->coro_ref = luaL_ref(tarantool_L, LUA_REGISTRYINDEX);
+	} else {
+		ctx->L = fiber()->storage.lua.stack;
+		ctx->coro_ref = LUA_REFNIL;
+	}
+	ctx->idx = 0;
+	ctx->top_svp = lua_gettop(ctx->L);
+	lua_rawgeti(ctx->L, LUA_REGISTRYINDEX, func->func_ref);
+}
+
+/**
+ * Releases Lua state.
+ */
+static void
+lua_func_adapter_end(struct func_adapter_ctx *base)
+{
+	struct lua_func_adapter_ctx *ctx = (struct lua_func_adapter_ctx *)base;
+	lua_settop(ctx->L, ctx->top_svp);
+	luaL_unref(tarantool_L, LUA_REGISTRYINDEX, ctx->coro_ref);
+	ctx->coro_ref = LUA_REFNIL;
+	ctx->L = NULL;
+}
+
+/**
+ * Call the function with arguments that were passed before.
+ */
+static int
+lua_func_adapter_call(struct func_adapter_ctx *base_ctx)
+{
+	struct lua_func_adapter_ctx *ctx =
+		(struct lua_func_adapter_ctx *)base_ctx;
+	int nargs = lua_gettop(ctx->L) - ctx->top_svp - 1;
+	if (luaT_call(ctx->L, nargs, LUA_MULTRET) != 0)
+		return -1;
+	ctx->idx = ctx->top_svp + 1;
+	return 0;
+}
+
+static void
+lua_func_adapter_push_number(struct func_adapter_ctx *base, double val)
+{
+	struct lua_func_adapter_ctx *ctx = (struct lua_func_adapter_ctx *)base;
+	lua_pushnumber(ctx->L, val);
+}
+
+static void
+lua_func_adapter_push_integer(struct func_adapter_ctx *base, int64_t val)
+{
+	struct lua_func_adapter_ctx *ctx = (struct lua_func_adapter_ctx *)base;
+	lua_pushinteger(ctx->L, val);
+}
+
+static void
+lua_func_adapter_push_string(struct func_adapter_ctx *base,
+			     const char *str, size_t len)
+{
+	struct lua_func_adapter_ctx *ctx = (struct lua_func_adapter_ctx *)base;
+	lua_pushlstring(ctx->L, str, len);
+}
+
+static void
+lua_func_adapter_push_tuple(struct func_adapter_ctx *base, struct tuple *tuple)
+{
+	struct lua_func_adapter_ctx *ctx = (struct lua_func_adapter_ctx *)base;
+	luaT_pushtuple(ctx->L, tuple);
+}
+
+static void
+lua_func_adapter_push_null(struct func_adapter_ctx *base)
+{
+	struct lua_func_adapter_ctx *ctx = (struct lua_func_adapter_ctx *)base;
+	lua_pushnil(ctx->L);
+}
+
+static bool
+lua_func_adapter_is_number(struct func_adapter_ctx *base)
+{
+	struct lua_func_adapter_ctx *ctx = (struct lua_func_adapter_ctx *)base;
+	return lua_gettop(ctx->L) >= ctx->idx && lua_isnumber(ctx->L, ctx->idx);
+}
+
+static void
+lua_func_adapter_pop_number(struct func_adapter_ctx *base, double *out)
+{
+	struct lua_func_adapter_ctx *ctx = (struct lua_func_adapter_ctx *)base;
+	assert(lua_isnumber(ctx->L, ctx->idx));
+	*out = lua_tonumber(ctx->L, ctx->idx++);
+}
+
+static bool
+lua_func_adapter_is_integer(struct func_adapter_ctx *base)
+{
+	/* There is no integer type in Lua. */
```

Comment:
There's int64 type in Tarantool. Also, we could return true if the Lua number can be converted to `int64_t` and false otherwise.

Reply:
Discussed f2f, decided not to imlement this functionality.